added valid_move? method to king

shaunshapiro opened the pull request.

bluem00se committed:
added valid_move? method to king

shaunshapiro commented:
I'm not sure if I did this as intended. I thought at first I would have to use the is_obstructed? method since we said we were building off it with this one, but is_obstructedd checks for the pieces inbetween the start and end, and the king only has a start and end since it moves only one space. I did try to follow @tzinser 's method to create something similar, though. @johnnyiller what do you think?

johnnyiller commented on app/models/king.rb:
```diff
@@ -1,4 +1,24 @@
+
+		#determine if move is only 1 space in any direction
+		if (start_x - end_x).abs <= 1 && (start_y - end_y).abs <= 1
+			#then check if space is occupied
```
I don't think we want to check if the space is occupied, just that the move is a valid one.

johnnyiller replied:
so I think if just 

return (start_x - end_x).abs <= 1 && (start_y - end_y).abs <= 1

you should be all set.

johnnyiller commented on app/models/king.rb:
```diff
@@ -1,4 +1,24 @@
 class King < Piece
-	def move
+
```
It's probably okay to skip it for this app but in a real world application you would want to actually validate that the co-ordinates that the user wants to move to are valid in the first place.  For example a user could put in a start or end point that was off the board and this method would not detect it.  I'm not saying that this has to be done for this app but just pointing out what would need to be done for a public production app.

johnnyiller replied:
additionally in a production app you'd want to make sure that the piece was not being told to move to the same location it was already on.  So start_point != end_point

johnnyiller commented on app/models/king.rb:
```diff
@@ -1,4 +1,24 @@
+			
+			#once movement for other pieces is sorted out, this method will also have to check
+			#if the space the king is moving into will put it in check, which will also be an
+			#invalid move
```
yes

johnnyiller commented:
also no need to use the is_obstructed in this method if that's an impossibility btw so you did the right thing with that.

bluem00se committed:
update valid_move to check board coordinates, that the piece is moved to a diff location, & removed check to see if a location is occupied.

shaunshapiro commented:
Got it. I updated the file per your suggestions. Figured it wouldn't be too hard to check if the move was on the board so I added that check as well.

johnnyiller commented on app/models/king.rb:
```diff
@@ -1,4 +1,27 @@
 class King < Piece
-	def move
+
+	def valid_move?(start_point, end_point)
```
Looking at this, it doesn't make sense to pass in the start_point.  If you are already dealing with a piece on the board then you already know the start point.  This is a bit advanced.  But look at the code we have here.  There are several things that should be methods in the Piece model so that all pieces can use them.  
1. getting the start position
2. checking the move is on the board
3. checking the start != end point

you should extract the pieces of code you are likely to re-use and put them in the Piece parent class.  Then you could rewrite this method like the following.

def valid_move?(end_point)
   return false unless self.on_board?(end_point)
   return false unless self.new_point?(end_point)
   (self. pos_x - end_x).abs <= 1 && (self. pos_y - end_y).abs <= 1
end

johnnyiller commented:
I wanted to say you can if you wish push this into production and re-factor at a later point in time if you wish.  What you have is working correctly so I wouldn't block next weeks work just to re-factor the code a little bit.

tzinser commented:
@johnnyiller can you please check trello regarding the capture logic question I have posted.

shaunshapiro commented:
@johnnyiller I'll try to refactor it today if I can but good to know. I'm having some trouble on the other task, listing out the games with a scope and passing that to the controller.
